tests/test_bingham.py

```diff
@@ -0,0 +1,194 @@
+                "youngs_modulus": 1.0e7,
+                "poisson_ratio": 0.3,
+                "tau0": 200.0,
+                "mu": 200.0,
+                "critical_shear_rate": 0.2,
+                "ndim": 2,
+            }
+        ),
+        None,
+        Particles(
+            jnp.array([[0.5, 0.5]]).reshape(1, 1, 2),
+            (
+                Bingham(
+                    {
+                        "density": 1000,
+                        "youngs_modulus": 1.0e7,
+                        "poisson_ratio": 0.3,
+                        "tau0": 200.0,
+                        "mu": 200.0,
+                        "critical_shear_rate": 0.2,
+                        "ndim": 2,
+                    }
+                )
+            ),
+            jnp.array([0]),
+        ),
+        {"pressure": jnp.zeros(1)},
+        Quadrilateral4Node(
+            (1, 1),
+            1,
+            (4.0, 4.0),
+            [(0, Constraint(0, 2.0)), (0, Constraint(1, 3.0))],
+            Nodes(4, jnp.array([-2, -2, 2, -2, -2, 2, 2, 2]).reshape((4, 1, 2))),
+        ),
+        jnp.array(
+            [-5208520.35574006, -5208613.86694342, 0.0, -233.778008402801, 0.0, 0.0]
+        ).reshape((6, 1)),
+        1.0,
+    ),
+    (
+        Bingham(
+            {
+                "density": 1000,
+                "youngs_modulus": 1.0e7,
+                "poisson_ratio": 0.3,
+                "tau0": 200.0,
+                "mu": 200.0,
+                "critical_shear_rate": 0.2,
+                "ndim": 2,
+                "incompressible": True,
+            }
+        ),
+        None,
+        Particles(
+            jnp.array([[0.5, 0.5]]).reshape(1, 1, 2),
+            (
+                Bingham(
+                    {
+                        "density": 1000,
+                        "youngs_modulus": 1.0e7,
+                        "poisson_ratio": 0.3,
+                        "tau0": 200.0,
+                        "mu": 200.0,
+                        "critical_shear_rate": 0.2,
+                        "ndim": 2,
+                        "incompressible": True,
+                    }
+                )
+            ),
+            jnp.array([0]),
+        ),
+        {"pressure": jnp.zeros(1)},
+        Quadrilateral4Node(
+            (1, 1),
+            1,
+            (4.0, 4.0),
+            [(0, Constraint(0, 2.0)), (0, Constraint(1, 3.0))],
+            Nodes(4, jnp.array([-2, -2, 2, -2, -2, 2, 2, 2]).reshape((4, 1, 2))),
+        ),
+        jnp.array(
+            [-187.0224067222, -280.5336100834, 0.0, -233.778008402801, 0.0, 0.0]
+        ).reshape((6, 1)),
+        1.0,
+    ),
+]
+
+
+@pytest.mark.parametrize(
+    "material, dstrain, particles, state_vars, element, target, dt",
+    material_dstrain_particles_state_vars_element_targets_dt,
+)
+def test_compute_stress(material, dstrain, particles, state_vars, element, target, dt):
+    if element.constraints != []:
+        element.apply_boundary_constraints()
```

Comment:
```suggestion
    particles.update_natural_coords(element)
    if element.constraints:
        element.apply_boundary_constraints()
```
The natural coordinates of the particles would need to be updated based on the elements the particles are in. Since this was not happening, the gradient of the shape function was being calculated incorrectly thereby giving wrong strain values. The reason volumetric strain worked was because volumetric strain is calculated at centroid, which has the same natural coordinates always. Making this change will pass all the test cases.

Also, as a note, you should be able to clean up the test cases quite a bit. You don't need to create a separate material variable if you have the particles. You should just use the material of the particles. `element`, `particles` and `state_vars` can be moved to fixtures, instead of declaring it in each test repeatedly.